Mathlib/Probability/Hoeffding.lean

```diff
@@ -0,0 +1,556 @@
+/-
+Copyright (c) 2024 Kei Tsukamoto. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kei Tsukamoto, Kazumi Kasaura, Naoto Onda, Sho sonoda, Yuma Mizuno
+-/
+
+import Mathlib.Analysis.Calculus.ParametricIntegral
+import Mathlib.Analysis.SpecialFunctions.Log.Deriv
+import Mathlib.MeasureTheory.Measure.Tilted
+import Mathlib.Probability.Variance
+
+/-!
+# Hoeffding's lemma
+
+This file states Hoeffding's lemma. We introduce cumulant to complete the proof.
+
+## Main results
+
+* `ProbabilityTheory.tilt_first_deriv`: derivation of `μ[exp (t * X ω)]` is
+  `μ[exp (t * X ω) * X ω]`. In order to deal with the differentiation of parametric integrals,
+`hasDerivAt_integral_of_dominated_loc_of_deriv_le` are used in the proof.
+* `ProbabilityTheory.tilt_second_deriv`: derivation of `μ[fun ω ↦ rexp (t * X ω) * X ω]` is
+  `μ[fun ω ↦ rexp (t * X ω) * X ω ^ 2]`. In order to deal with the differentiation of
+  parametric integrals, `hasDerivAt_integral_of_dominated_loc_of_deriv_le` are used in the proof.
+* `ProbabilityTheory.cum_deriv_one`: first derivative of cumulant `log μ[rexp (t * X ω))]`.
+  It can be described by exponential tilting.
+* `ProbabilityTheory.cum_deriv_two`: second derivative of cumulant `log μ[rexp (t * X ω))]`.
+* `ProbabilityTheory.hoeffding`: Hoeffding's Lemma states that for a random variable `X` with
+  `E[X] = 0` (zero mean) and `a ≤ X ≤ b` almost surely, the inequality
+  `μ[exp (t * (X ω))] ≤ exp (t^2 * (b - a)^2 / 8)` holds almost surely for all `t ∈ ℝ`.
+
+## References
+
+* Martin J. Wainwright. High-Dimensional Statistics. Cambridge university press, 2019. Chapter 2
+* Mehryar Mohri. Foundations of Machine Learning. The MIT Press, 2018. Chapter D
```

Comment:
See https://leanprover-community.github.io/contribute/doc.html#citing-other-works for how to properly cite other works in Mathlib.